storage-access.bs

```diff
@@ -4,27 +4,111 @@ Shortname: storage-access
+
+<h2 id="the-storage-access-api">The Storage Access API</h2>
+
+Each {{Document}} has an associated <dfn export for=Document id=storage-access-flag>has storage access flag</dfn>, initially unset.
+
+<pre class="idl">
+partial interface Document {
+  Promise&lt;boolean> hasStorageAccess();
+  Promise&lt;void> requestStorageAccess();
+};
+</pre>
+
+When invoked, |document|.<dfn export method for=Document><code>hasStorageAccess()</code></dfn> must run these steps:
+
+1. Let |p| be [=a new promise=].
+1. If |document|'s [=has storage access flag=] is set, [=resolve=] |p| with true and abort these steps.
+1. If the |document|'s [=active sandboxing flag set=] has its [=sandboxed origin browsing context flag=] set, [=resolve=] |p| with false and abort these steps.
+1. If |document|'s <a for=Document>browsing context</a> is a [=top-level browsing context=], [=resolve=] |p| with true and abort these steps.
+1. If |document|'s <a for=Document>browsing context</a> is [=same origin=] with its [=top-level browsing context=], [=resolve=] |p| with true and abort these steps.
+1. <span class=XXX>Finish this algorithm.</span>
+1. [=If aborted=], return |p|.
+
+<div class=issue>
+<p>Complete this algorithm.
+<p>See its <a href="https://developer.mozilla.org/en-US/docs/Web/API/Document/hasStorageAccess">MDN page</a>, <a href="https://trac.webkit.org/browser/webkit/trunk/Source/WebCore/dom/DocumentStorageAccess.cpp#L80">WebKit implementation</a>, and <a href="https://hg.mozilla.org/mozilla-central/file/tip/dom/base/Document.cpp#l15558">Gecko implementation</a>.
+</ul>
+</div>
+
+When invoked, |document|.<dfn export method for=Document><code>requestStorageAccess()</code></dfn> must run these steps:
+
+1. Let |p| be [=a new promise=].
+1. If |document|'s [=has storage access flag=] is set, [=resolve=] |p| and abort these steps.
+1. If the |document|'s [=active sandboxing flag set=] has its [=sandboxed origin browsing context flag=] set, [=reject=] |p| and abort these steps.
+1. If |document|'s [=has storage access flag=] is unset, [=reject=] |p| and abort these steps.
+1. If |document|'s <a for=Document>browsing context</a> is a [=top-level browsing context=], [=resolve=] |p| and abort these steps.
+1. If |document|'s <a for=Document>browsing context</a> is [=same origin=] with its [=top-level browsing context=], [=resolve=] |p| and abort these steps.
+1. If the |document|'s {{Window}} object has [=transient activation=] and the |document|'s [=active sandboxing flag set=] has its [=sandbox storage access by user activation flag=] set, [=reject=] |p| and abort these steps.
+1. If |document|'s <a for=Document>browsing context</a>'s [=opener browsing context=] is not its [=top-level browsing context=], [=reject=] |p| and abort these steps.
+1. If the algorithm is invoked when the |document|'s {{Window}} object does not have [=transient activation=], [=reject=] |p| and abort these steps.
+1. <span class=XXX>Finish this algorithm.</span>
```

Comment:
Something to consider is to add preserving the user gesture in all cases but explicit "Don't allow" from the user. That's what WebKit does to allow the iframe to do a popup if it turns out that the user is not logged in yet.

Reply:
I think I've done this in 8c9b716.